### ISSUE TITLE
Installation and execution of NetVips

### ISSUE THREAD

**Issue by roganjoshua** [kleisauke/net-vips#3]
Hi I am trying to get netvips working here on windows.Will this work?

Do I need to set up Environment Variables first?

I get this

ibvips zip file not in download directory. Downloading now ...
Downloading file: https://github.com/jcupitt/libvips/releases/download/v/vips-.zip

which of course fails

All the tests fail.

Not sure where to put the downloaded vips folder? 

  System.TypeInitializationException : The type initializer for '<Module>' threw an exception.
[02/05/2018 13:30:36 Informational] [xUnit.net 00:00:01.3574786]       ---- System.DllNotFoundException : Unable to load DLL 'libvips-42.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
[02/05/2018 13:30:36 Informational] [xUnit.net 00:00:01.3575698]       Stack Trace:
[02/05/2018 13:30:36 Informational] [xUnit.net 00:00:01.3576566]            at NetVips.Tests.TestsFixture..ctor()
[02/05/2018 13:30:36 Informational] [xUnit.net 00:00:01.3577459]         ----- Inner Stack Trace -----
[02/05/2018 13:30:36 Informational] [xUnit.net 00:00:01.3578316]            at NetVips.Internal.Vips.VipsInit(String argv0)
[02/05/2018 13:30:36 Informational] [xUnit.net 00:00:01.3579297]         C:\Users\Martin_Sarosi\Downloads\net-vips-master\src\NetVips\ModuleInitializer.cs(13,0): at NetVips.ModuleInitializer.Initialize()
[02/05/2018 13:30:36 Informational] [xUnit.net 00:00:01.3580023]            at .cctor()

Any help

**Comment by kleisauke** [comment]
Hello @roganjoshua,

It looks like you tried to run `.\build.ps1`. When running the Cake Build script a few environment needs to be set first. It should work like this in PowerShell:
```ps1
$env:VIPS_VERSION = "8.6.3"
$env:VIPS_ZIP_VERSION = "dev-w64-all-8.6.3-1"
$env:VIPS_HOME = "C:\vips-dev-$env:VIPS_VERSION"
$env:PATH = "$env:VIPS_HOME\bin;$env:PATH"
.\build.ps1
```

However, the Cake Build script is only used for the continuous integration builds, so it's not needed to run it. When you want to use NetVips in your project(s) I can recommend installing it with NuGet. In the NuGet Package Manager Console it can be installed like this:
```bash
Install-Package NetVips
```

You'll only need the libvips shared library on your library search path, see for more info:
https://jcupitt.github.io/libvips/install.html#installing-the-windows-binary

**Comment by roganjoshua** [comment]
Thanks for such a quick reply. My problem is finding the vips dlls. 

On windows what is the library search path?

I don't where to put it for the tests to run.

Apologies for my idiocy.

NetVips.Tests.ArithmeticTests.TestAbs [FAIL]
[03/05/2018 07:32:04 Informational] [xUnit.net 00:00:01.0535451]       System.TypeInitializationException : The type initializer for '<Module>' threw an exception.
[03/05/2018 07:32:04 Informational] [xUnit.net 00:00:01.0538460]       ---- System.DllNotFoundException : Unable to load DLL 'libvips-42.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
[03/05/2018 07:32:04 Informational] [xUnit.net 00:00:01.0553459]       Stack Trace:
[03/05/2018 07:32:04 Informational] [xUnit.net 00:00:01.0571805]            at NetVips.Tests.TestsFixture

**Comment by roganjoshua** [comment]
I copied them into Test bin\debug folder and now get 

System.TypeInitializationException : The type initializer for '<Module>' threw an exception.
[03/05/2018 07:50:21 Informational] [xUnit.net 00:00:00.8170957]       ---- System.BadImageFormatException : An attempt was made to load a program with an incorrect format. (Exception from HRESULT: 0x8007000B)
[03/05/2018 07:50:21 Informational] [xUnit.net 00:00:00.8181482]       Stack Trace:

**Comment by RyanYANG52** [comment]
I think 'TypeInitializationException' is because you are using 32bit, and libvips bin is win64, mine works fine. Just use x64 instead of AnyCpu @roganjoshua

**Comment by kleisauke** [comment]
Ah, I overlooked the `BadImageFormatException`, the recommendation from @RyanYANG52 should work in that case.

**Comment by jcupitt** [comment]
The libvips binary `vips-dev-w64-all-x.y.z.zip` contains ALL the file format loaders that libvips supports, including some very minor ones, and a complete copy of ImageMagick. It can process many different file types, but it is also rather vulnerable to hacking. I'm sure it would be easy to exploit with crafted input files.

The `vips-dev-w64-web-x.y.z.zip` build contains only file format loaders that have been carefully tested and which are commonly used on the web. It won't load formats like openslide (for example), but it should be much safer to use with untrusted input images.

**Comment by NiteshSaxena** [comment]
@RyanYANG52 , Many thanks, saved the day!.

**Comment by kleisauke** [comment]
I've been thinking whether there is a better solution to prevent users from downloading the libvips binaries and the need to set their `PATH` environment variable to find the required DLL's on Windows.

I could set-up the NuGet package to bundle the latest pre-compiled libvips binaries (this can be done through making a `runtimes` folder which has sub-folders named `{platform}-{architecture}\lib\native`). Magick.NET solves the problem from setting the `PATH` environment and the need to download the ImageMagick binary in that way.

Unfortunately there a disadvantages to this approach:
- Users are unable to choose a different version of libvips because the libvips binary files are included in the NuGet package.
- Maintaining NetVips becomes more difficult because for every libvips release, NetVips also needs to be updated (I now only need to update NetVips if there are API changes in libvips).

The advantage of this approach is that users can install NetVips easier without the need to set-up libvips first.

I weigh the pros and cons and I think we have to stick to the current installation / packaging behavior. Only on Windows it's slightly more difficult to install. On MacOS libvips is on Homebrew, MacPorts and Fink. On Linux it's in included in the package manager (for example Debian, Ubuntu, CentOS, RHEL, Fedora, Alpine, SuSE, FreeBSD, etc. have libvips in their package manager).

For Windows, I think vcpkg can help with this in the future, see these open issues:
https://github.com/jcupitt/libvips/issues/868
https://github.com/Microsoft/vcpkg/issues/2676

**Comment by jcupitt** [comment]
Yes, a NuGet package that installed libvips would be useful. Could NetVips depend on it and have it pulled in automatically?

**Comment by kleisauke** [comment]
Merged to master. I will try to make a new release within a few days.

Making one DLL of libvips would be ideal, but I'm not sure how difficult this is. I'll investigate whether it is possible to do this in https://github.com/jcupitt/build-win64.

**Comment by jcupitt** [comment]
Could a merged DLL be a licence violation? The LGPL says users need to be able to replace shared libraries, and libvips has a lot of LGPL dependencies.

**Comment by kleisauke** [comment]
I've created https://github.com/jcupitt/build-win64/issues/21 in order to keep this issue clean.

**Comment by kleisauke** [comment]
I've just released version 1.0.3 on NuGet. The pre-compiled libvips Windows binary is bundled with this version. It's therefore no longer necessary to download the pre-compiled binary and to set the `PATH` environment variable.

The readme has some remarks about how to disable the bundled libvips and use a different version of libvips.

I'm sorry it took so long, I had some busy weeks on my university. I'll close this issue for now, please feel free to re-open if there's still a problem.

**Comment by JPomichael** [comment]
> I think 'TypeInitializationException' is because you are using 32bit, and libvips bin is win64, mine works fine. Just use x64 instead of AnyCpu @roganjoshua

wow. successed. thx
